pyocd/rtos/threadx.py

```diff
@@ -0,0 +1,514 @@
+        super(ThreadXThreadProvider, self).__init__(target)
+        self._created_ptr = None
+        self._created_cnt = None
+        self._current_ptr = None
+        self._system_state = None
+        self._threads = {}
+
+    def init(self, symbolProvider):
+        self._created_ptr = symbolProvider.get_symbol_value(
+            '_tx_thread_created_ptr')
+        if self._created_ptr is None:
+            return False
+        LOG.debug("ThreadX: _tx_thread_created_ptr = 0x%08x", self._created_ptr)
+
+        self._created_cnt = symbolProvider.get_symbol_value(
+            '_tx_thread_created_count')
+        if self._created_cnt is None:
+            return False
+        LOG.debug("ThreadX: _tx_thread_created_cnt = 0x%08x", self._created_cnt)
+
+        self._current_ptr = symbolProvider.get_symbol_value(
+            '_tx_thread_current_ptr')
+        if self._current_ptr is None:
+            return False
+        LOG.debug("ThreadX: _tx_thread_current_ptr = 0x%08x", self._current_ptr)
+
+        self._system_state = symbolProvider.get_symbol_value(
+            '_tx_thread_system_state')
+        if self._system_state is None:
+            return False
+        LOG.debug("ThreadX: _tx_thread_system_state = 0x%08x",
+                  self._current_ptr)
+
+        self._target.session.subscribe(
+            self.event_handler, Target.Event.POST_FLASH_PROGRAM)
+        self._target.session.subscribe(
+            self.event_handler, Target.Event.POST_RESET)
+
+        return True
+
+    def invalidate(self):
+        self._threads = {}
+
+    def event_handler(self, notification):
+        # Invalidate threads list if flash is reprogrammed.
+        LOG.debug("ThreadX: invalidating threads list: %s" %
+                  (repr(notification)))
+        self.invalidate()
+
+    def _build_thread_list(self):
+        # Read the number of threads.
+        threadCount = self._target_context.read32(self._created_cnt)
+
+        # Build up list of all the threads
+        allThreads = TargetList(self._target_context, self._created_ptr)
+        newThreads = {}
+        for threadBase in allThreads:
+            try:
+                # Reuse existing thread objects if possible.
+                if threadBase in self._threads:
+                    t = self._threads[threadBase]
+
+                    # Ask the thread object to update its state and priority.
+                    t.update_info()
+                else:
+                    t = ThreadXThread(self._target_context, self, threadBase)
+                LOG.debug("Thread 0x%08x (%s)", threadBase, t.name)
+                newThreads[t.unique_id] = t
+            except exceptions.TransferError:
+                LOG.debug(
+                    "TransferError while examining thread 0x%08x", threadBase)
+
+        # Is the number of threads correct?
+        if len(newThreads) != threadCount:
+            LOG.warning("ThreadX: thread count mismatch, expected: {0}, found: {1}".format(
```

Comment:
Use `%` formatting instead of `format()` method, like the other log calls. The logger stores the string formatting arguments in the log message object and only formats the string if it gets printed.

Reply:
Agreed. Did not know it would make a difference.